File: ruby/test/task_master/ascii_table_test.rb
require_relative "../test_helper"

class TaskMaster::AsciiTableTest < Minitest::Test
  def setup
    @headers = ["Header 1", "Header 2"]
    @rows = [["foo", "1"], ["bar", "12"]]
    @class_name = TaskMaster::AsciiTable
  end

  def test_should_write_table_to_stdout
    output = capture_io do
      @class_name.new(@headers, @rows).draw
    end
    assert_equal(expected_table, output.join)
  end

  def test_should_return_table_where_each_line_of_the_table_has_same_number_characters
    output = capture_io do
      @class_name.new(@headers, @rows).draw
    end
    assert(
      output.join.split("\n").map(&:length).uniq.length == 1
    )
  end

  def test_should_dynamically_resize_table_based_on_cell_contents_size
    different_sized_headers = ["Extra long header 1", "Extra long header 2"]
    output_one = capture_io do
      @class_name.new(@headers, @rows).draw
    end
    output_two = capture_io do
      @class_name.new(different_sized_headers, @rows).draw
    end

    refute_equal(
      output_one.join.split("\n").map(&:length).uniq.first,
      output_two.join.split("\n").map(&:length).uniq.first
    )
  end

  def expected_table
    [
      "+----------+----------+\n",
      "| Header 1 | Header 2 |\n",
      "+----------+----------+\n",
      "| foo      | 1        |\n",
      "| bar      | 12       |\n",
      "+----------+----------+\n"
    ].join
  end
end

File: javascript/test/AsciiTable.test.js
const path = require('path');
const AsciiTable = require(path.join(__dirname, '../lib/AsciiTable'));
const originalConsoleLog = console.log;
let logOutput = [];

beforeAll(() => {
	console.log = jest.fn((...args) => {
		logOutput.push(args.join(' '));
	});
	process.stdout.write = jest.fn((...args) => {
		logOutput.push(args.join(' '));
	});
});

afterAll(() => {
	console.log = originalConsoleLog;
});

beforeEach(() => {
	headers = ["Header 1", "Header 2"];
	rows = [["foo", "1"], ["bar", "12"]];
	class_name = AsciiTable;
});

test("should write table to stdout", () => {
	new class_name(headers, rows).draw();
	expect(logOutput.join('')).toContain(expectedTable());
});

test("should return table where each line of the table has same number of characters", () => {
	new class_name(headers, rows).draw();
	expect(
		Array.from(
			new Set(
				logOutput.join('').split("\n").map((line) => {
					if(line.lenth > 0) {
						return line.length
					}
				})
			)
		).length
	).toEqual(1)
});

test("should dynamically resize table based on cell contents size", () => {
	differentSizedHeaders = ["Extra long header 1", "Extra long header 2"]
	new class_name(headers, rows).draw()
	const outputOne =
		Array.from(
			new Set(
				logOutput.join('').split("\n").map((line) => {
					return line.length
				})
			)
		)[0]
	console.log = originalConsoleLog;
	logOutput.length = 0
	console.log = jest.fn((...args) => {
		logOutput.push(args.join(' '));
	});
	new class_name(differentSizedHeaders, rows).draw()
	console.log = originalConsoleLog;
	const outputTwo =
		Array.from(
			new Set(
				logOutput.join('').split("\n").map((line) => {
					return line.length
				})
			)
		)[0]
	expect(outputOne).not.toEqual(outputTwo)
})

function expectedTable() {
	return [
		"+----------+----------+",
		"| Header 1 | Header 2 |",
		"+----------+----------+",
		"| foo      | 1        |",
		"| bar      | 12       |",
		"+----------+----------+"
	].join("")
}
